src/scripts/main.js

```diff
@@ -1,3 +1,202 @@
+}
+
+function handleInput(e) {
+  switch (e.key) {
+    case 'ArrowUp':
+      if (!canMoveUp()) {
+        setupInputOnce();
+
+        return;
+      }
+      moveUp();
+      break;
+
+    case 'ArrowDown':
+      if (!canMoveDown()) {
+        setupInputOnce();
+
+        return;
+      }
+      moveDown();
+      break;
+
+    case 'ArrowLeft':
+      if (!canMoveLeft()) {
+        setupInputOnce();
+
+        return;
+      }
+      moveLeft();
+      break;
+
+    case 'ArrowRight':
+      if (!canMoveRight()) {
+        setupInputOnce();
+
+        return;
+      }
+      moveRight();
+      break;
+
+    default:
+      setupInputOnce();
+
+      return;
+  }
+
+  const newTile = new Tile(gameField);
+
+  grid.getRandomEmptyCell().linkTile(newTile);
+
+  if (!canMoveUp() && !canMoveDown() && !canMoveLeft() && !canMoveRight()) {
+    document.querySelector('.message-lose').classList.remove('hidden');
+  }
+
+  setupInputOnce();
+}
+
+function moveUp() {
+  slideTiles(grid.cellsGroupedByColumn);
+}
+
+function moveDown() {
+  slideTiles(grid.cellsGroupedByReversedColumn);
+}
+
+function moveLeft() {
+  slideTiles(grid.cellsGroupedByRow);
+}
+
+function moveRight() {
+  slideTiles(grid.cellsGroupedByReversedRow);
+}
+
+function slideTilesInGroup(group) {
+  for (let i = 0; i < group.length; i++) {
+    if (group[i].isEmpty()) {
+      continue;
+    }
+
+    const cellWithTail = group[i];
+
+    let targetCell;
+    let j = i - 1;
+
+    while (j >= 0 && group[j].canAccept(cellWithTail.linkedTile)) {
+      targetCell = group[j];
+      j--;
+    }
+
+    if (!targetCell) {
+      continue;
+    }
+
+    if (targetCell.isEmpty()) {
+      targetCell.linkTile(cellWithTail.linkedTile);
+    } else {
+      targetCell.linkTileForMerge(cellWithTail.linkedTile);
+    }
+
+    cellWithTail.unlinkTile();
+  }
+}
+
+function slideTiles(groupedCells) {
+  groupedCells.forEach(group => slideTilesInGroup(group));
+
+  grid.cells.forEach(cell => {
+    if (cell.hasTileForMerge()) {
+      cell.mergeTiles();
+      score += cell.getValue();
+      scoreElement.innerHTML = score;
+    }
+
+    if (cell.getValue() === 2048) {
+      document.querySelector('.message-win').classList.remove('hidden');
+      window.removeEventListener('keydown');
+    }
+  });
+}
+
+function setupInputOnce() {
+  window.addEventListener('keydown', handleInput, { once: true });
+};
+
+function canMoveUp() {
+  return canMove(grid.cellsGroupedByColumn);
+}
+
+function canMoveDown() {
+  return canMove(grid.cellsGroupedByReversedColumn);
+}
+
+function canMoveLeft() {
+  return canMove(grid.cellsGroupedByRow);
+}
+
+function canMoveRight() {
+  return canMove(grid.cellsGroupedByReversedRow);
+}
+
+function canMove(groupedCells) {
+  return groupedCells.some(group => canMoveInGroup(group));
+}
+
+function canMoveInGroup(group) {
+  return group.some((cell, index) => {
+    if (index === 0) {
+      return false;
+    }
+
+    if (cell.isEmpty()) {
+      return false;
+    }
```

Comment:
Combine similar conditions
```suggestion
    if (!index || cell.isEmpty()) {
      return false;
    }
```